internal/serving/grpcserver.go

```diff
@@ -104,72 +148,114 @@ func (gw *GrpcWrapper) Start() error {
+	httpMux.HandleFunc("/healthz", healthzServer(gw.grpcClient))
+	httpMux.HandleFunc("/swagger/", swaggerServer("."))
+	httpMux.Handle("/", proxyMux)
+	zpages.Handle(httpMux, "/debug")
+
+	gw.proxy = &http.Server{Handler: httpMux}
+	gw.proxyAwaiter = make(chan error)
+
 	go func() {
 		gw.logger.Infof("Serving proxy on :%d", gw.proxyLh.Number())
-		err := gw.proxy.ListenAndServe()
+		err := gw.proxy.Serve(proxyLn)
 		gw.proxyAwaiter <- err
 		if err != nil {
-			gw.logger.WithFields(log.Fields{"error": err.Error()}).Error("proxy ListenAndServe() error")
+			gw.logger.WithFields(log.Fields{
+				"error":           err.Error(),
+				"serviceEndpoint": serviceEndpoint,
+				"proxyPort":       gw.proxyLh.Number(),
+			}).Error("proxy ListenAndServe() error")
 		}
 	}()
 
 	return nil
 }
 
-// WaitForTermination blocks until the gRPC server is shutdown.
-func (gw *GrpcWrapper) WaitForTermination() (error, error) {
-	var grpcErr, proxyErr error
-
-	if gw.grpcAwaiter != nil {
-		grpcErr = <-gw.grpcAwaiter
-		close(gw.grpcAwaiter)
-		gw.grpcAwaiter = nil
+func (gw *GrpcWrapper) startGrpcServer() error {
+	// Starting gRPC server
+	if gw.serviceLn != nil {
+		return nil
 	}
 
-	if gw.proxyAwaiter != nil {
-		proxyErr = <-gw.proxyAwaiter
-		close(gw.proxyAwaiter)
-		gw.proxyAwaiter = nil
+	serviceLn, err := gw.serviceLh.Obtain()
+	if err != nil {
+		gw.logger.WithFields(log.Fields{
+			"error":       err.Error(),
+			"servicePort": gw.serviceLh.Number(),
+		}).Error("net.Listen() error")
+		return err
 	}
-	return grpcErr, proxyErr
-}
+	gw.serviceLn = serviceLn
 
-// Stop gracefully shutsdown the gRPC server.
-func (gw *GrpcWrapper) Stop() error {
-	if gw.server == nil {
-		return nil
-	}
-	gw.server.GracefulStop()
-	gw.proxy.Shutdown(context.Background())
+	gw.logger.WithFields(log.Fields{
+		"servicePort": gw.serviceLh.Number(),
+	}).Info("TCP net listener initialized")
 
-	portErr := gw.serviceLn.Close()
-	_ = gw.proxyLn.Close()
+	server := grpc.NewServer(grpc.StatsHandler(&ocgrpc.ServerHandler{}))
+	for _, handlerFunc := range gw.serviceHandlerFuncs {
+		handlerFunc(server)
+	}
+	gw.server = server
+	gw.grpcAwaiter = make(chan error)
 
-	grpcErr, proxyErr := gw.WaitForTermination()
+	go func() {
+		gw.logger.Infof("Serving gRPC on :%d", gw.serviceLh.Number())
+		err := gw.server.Serve(serviceLn)
+		gw.grpcAwaiter <- err
+		if err != nil {
+			gw.logger.WithFields(log.Fields{"error": err.Error()}).Error("gRPC serve() error")
+		}
+	}()
 
-	gw.server = nil
-	gw.serviceLn = nil
-	gw.proxy = nil
-	gw.proxyLn = nil
+	return nil
+}
 
-	if grpcErr != nil {
-		return grpcErr
+// healthzServer returns a simple health handler which returns ok.
+func healthzServer(conn *grpc.ClientConn) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "text/plain")
+		if s := conn.GetState(); s != connectivity.Ready {
+			http.Error(w, fmt.Sprintf("ClientConn is %s", s), http.StatusBadGateway)
```

Comment:
Because /healthz is querying general server health this should report 503 - ServiceUnavailable